Google.Api.Generator.Tests/ResourcePatternTest.cs

```diff
@@ -21,21 +21,113 @@ namespace Google.Api.Generator.Tests
-        public void ValidPattern(string pattern, string[] paramNames, string pathTemplateString, string[] expandArgs, string expanded)
+            "as/{a=*}/bs/{b_1_b_2}/cs/{c_1_c_2}/{d=**}")]
+        public void ValidPattern(string pattern, string[] paramNames, string pathTemplateString)
         {
             var pat = new ResourcePattern(pattern);
             Assert.Equal(paramNames, pat.ParameterNames);
             Assert.Equal(pathTemplateString, pat.PathTemplateString);
+        }
+
+        [Theory]
+        [InlineData("as/{a}", new[] { "A" }, "as/A")]
+        [InlineData("as/{a=*}", new[] { "A" }, "as/A")]
+        [InlineData("as/{a=**}", new[] { "A/AA/AAA" }, "as/A/AA/AAA")]
+        [InlineData("as/{a=*}/bs/{b=*}/cs/{c=**}", new[] { "A", "B", "C/CC/CCC" }, "as/A/bs/B/cs/C/CC/CCC")]
+        [InlineData("as/{a_1}~{a_2}.{a_3}", new[] { "A1", "A2", "A3" }, "as/A1~A2.A3")]
+        [InlineData("as/{a_1=*}~{a_2=*}.{a_3=*}", new[] { "A1", "A2", "A3" }, "as/A1~A2.A3")]
+        [InlineData("as/{a=*}/bs/{b_1}-{b_2}/cs/{c_1=*}_{c_2}/{d=**}", new[] { "A", "B1", "B2", "C1", "C2", "D" }, "as/A/bs/B1-B2/cs/C1_C2/D")]
+        public void ValidExpansion(string pattern, string[] expandArgs, string expanded)
+        {
+            var pat = new ResourcePattern(pattern);
             Assert.Equal(expanded, pat.Expand(expandArgs));
         }
 
+        [Theory]
+        [InlineData("as/bs/cs", "as/bs/cs")]
+
+        [InlineData("*", "[^/]+")]
+        [InlineData("*/bs/cs", "[^/]+/bs/cs")]
+        [InlineData("as/*/cs", "as/[^/]+/cs")]
+        [InlineData("as/bs/*", "as/bs/[^/]+")]
+
+        [InlineData("{a=*}", "([^/]+)")]
+        [InlineData("{a=*}/bs/cs", "([^/]+)/bs/cs")]
+        [InlineData("{a=*/bs}/cs", "([^/]+/bs)/cs")]
+        [InlineData("as/{a=*}/cs", "as/([^/]+)/cs")]
+        [InlineData("as/{a=*/cs}", "as/([^/]+/cs)")]
+        [InlineData("as/bs/{a=*}", "as/bs/([^/]+)")]
+        [InlineData("as/{a=bs/*}", "as/(bs/[^/]+)")]
+
+        [InlineData("{a}", "([^/]+)")]
+        [InlineData("{a}/bs/cs", "([^/]+)/bs/cs")]
+        [InlineData("as/{a}/cs", "as/([^/]+)/cs")]
+        [InlineData("as/bs/{a}", "as/bs/([^/]+)")]
+
+        [InlineData("**", ".*")]
+        [InlineData("**/bs/cs", ".*/bs/cs")]
+        [InlineData("as/**/cs", "as(?:/.*)?/cs")]
+        [InlineData("as/bs/**", "as/bs(?:/.*)?")]
+
+        [InlineData("{a=**}", "(.+)")]
+        [InlineData("{a=**}/bs/cs", "(.+)/bs/cs")]
+        [InlineData("{a=**/bs}/cs", "(.*/bs)/cs")]
+        [InlineData("as/{a=**}/cs", "as/(.+)/cs")]
+        [InlineData("as/{a=**/cs}", "as/(.*/cs)")]
+        [InlineData("as/bs/{a=**}", "as/bs/(.+)")]
+        [InlineData("as/{a=bs/**}", "as/(bs(?:/.*)?)")]
+
+        [InlineData("as/*/bs/**", "as/[^/]+/bs(?:/.*)?")]
+        [InlineData("as/{a=*}/bs/{b=**}", "as/([^/]+)/bs/(.+)")]
+        [InlineData("{a=as/*}/{b=bs/**}", "(as/[^/]+)/(bs(?:/.*)?)")]
+        [InlineData("as/{a=*}/bs/{b=**}", "as/([^/]+)/bs/(.+)")]
```

Comment:
Not sure why it's giving this error twice, but it's right - there's a duplicate on line 91.

Reply:
Removed the duplicates.